### PR TITLE
Support nunit3

### LINKED ISSUE #98: Support NUnit 3
Currrently NUnit v3 is in Beta. I have the updates to support it in the my branch [jhamm/FsCheck](https://github.com/jhamm/FsCheck) in the support-nunit3 branch.

The FsCheck.NUnit.Addins project was not needed to support NUnit 3, so I removed it.

The branch is not ready to be merged into the main branch because it is still in beta and not all of the related projects, like [NUnit.Runners](https://www.nuget.org/packages/NUnit.Runners/), support NUnit 3 yet. For testing, I built and changed the references to pointed the references to the local copies of the dll's.

### PR DESCRIPTION
**Pull request by jhamm**
Here are the changes to support nunit3. This should fix #98.

The structure has some similarities as the XUnit Property.

If the result is exhausted, the test will be marked as Failed.

The failure messages are based on the recommendations on [Running Tests](https://fscheck.github.io/FsCheck/RunningTests.html) and the FsCheck XUnit Property project.

In the two example Nunit Test projects, there are now a total of 4 expected failing tests that end with _shouldFail. One of the tests is the exhausted test.

Here is the command that was used for testing this PR from the FsCheck directory:
`.\packages\NUnit.Console\tools\nunit-console.exe examples\FsCheck.NUnit.Examples\bin\debug\FsCheck.NUnit.Examples.dll examples\FsCheck.NUnit.CSharpExamples\bin\debug\FsCheck.NUnit.CSharpExamples.dll`

### CODE REVIEW

**Comment by jhamm** [comment]
I will take a look at fixing the nuget side of this over the weekend.

File: FsCheck.sln [hunk 1]
```diff
@@ -38,8 +38,6 @@ Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "FsCheck.Test", "tests\FsChe
 EndProject
 Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "FsCheck.NUnit", "src\FsCheck.NUnit\FsCheck.NUnit.fsproj", "{6B1B9FF5-8E6D-4461-9D77-CD1619403684}"
 EndProject
-Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "FsCheck.NUnit.Addin", "src\FsCheck.NUnit.Addin\FsCheck.NUnit.Addin.fsproj", "{BF24DE38-C529-4638-BE50-A937EA3E6876}"
-EndProject
```
Comment (kurtschelfthout) [inline FsCheck.sln]: Yay.

File: examples/FsCheck.NUnit.CSharpExamples/FsCheck.NUnit.CSharpExamples.csproj [hunk 2]
```diff
@@ -35,6 +35,7 @@
     <Prefer32Bit>false</Prefer32Bit>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="FSharp.Core, Version=4.3.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL" />
```
Comment (kurtschelfthout) [inline examples/FsCheck.NUnit.CSharpExamples/FsCheck.NUnit.CSharpExamples.csproj]: This is a dup.

File: examples/FsCheck.NUnit.CSharpExamples/FsCheckNunitAddin.cs [hunk 3]
```diff
@@ -1,16 +1,15 @@
...
+    //public class FsCheckNunitAddin : IAddin
+    //{
+    //    public bool Install(IExtensionHost host)
+    //    {
+    //        var tcBuilder = new FsCheckTestCaseBuilder();
+    //        host.GetExtensionPoint("TestCaseBuilders").Install(tcBuilder);
+    //        return true;
+    //    }
+    //}
 }
```
Comment (kurtschelfthout) [inline examples/FsCheck.NUnit.CSharpExamples/FsCheckNunitAddin.cs]: Please remove the empty file.

File: examples/FsCheck.XUnit.CSharpExamples/FsCheck.XUnit.CSharpExamples.csproj [hunk 4]
```diff
@@ -67,7 +67,7 @@
   -->
   <Import Project="..\..\.paket\paket.targets" />
   <Choose>
-    <When Condition="($(TargetFrameworkIdentifier) == '.NETFramework' And ($(TargetFrameworkVersion) == 'v2.0' Or $(TargetFrameworkVersion) == 'v3.0' Or $(TargetFrameworkVersion) == 'v3.5' Or $(TargetFrameworkVersion) == 'v4.0' Or $(TargetFrameworkVersion) == 'v4.5' Or $(TargetFrameworkVersion) == 'v4.5.1' Or $(TargetFrameworkVersion) == 'v4.5.2' Or $(TargetFrameworkVersion) == 'v4.5.3' Or $(TargetFrameworkVersion) == 'v4.6')) Or ($(TargetFrameworkIdentifier) == 'MonoAndroid') Or ($(TargetFrameworkIdentifier) == 'MonoTouch')">
+    <When Condition="$(TargetFrameworkIdentifier) == '.NETFramework' And ($(TargetFrameworkVersion) == 'v2.0' Or $(TargetFrameworkVersion) == 'v3.0' Or $(TargetFrameworkVersion) == 'v3.5' Or $(TargetFrameworkVersion) == 'v4.0' Or $(TargetFrameworkVersion) == 'v4.5' Or $(TargetFrameworkVersion) == 'v4.5.1' Or $(TargetFrameworkVersion) == 'v4.5.2' Or $(TargetFrameworkVersion) == 'v4.5.3' Or $(TargetFrameworkVersion) == 'v4.6')">
```
Comment (kurtschelfthout) [inline examples/FsCheck.XUnit.CSharpExamples/FsCheck.XUnit.CSharpExamples.csproj]: Is this change intentional? Doesn't seem related to NUnit.

File: paket.lock [hunk 5]
```diff
@@ -1,17 +1,18 @@
 NUGET
   remote: https://www.nuget.org/api/v2
   specs:
-    FAKE (4.0.1)
-    FSharp.Compiler.Service (1.3.1.0)
-    FSharp.Formatting (2.9.10)
+    FAKE (4.3.1)
+    FSharp.Compiler.Service (1.4.0.1)
+    FSharp.Formatting (2.10.0)
```
Comment (kurtschelfthout) [inline paket.lock]: Can you roll this back, I'll update dependencies on master at some point but prefer not to conflate.

File: paket.lock [hunk 6]
```diff
@@ -1,17 +1,18 @@
...
     FSharpVSPowerTools.Core (1.8.0)
       FSharp.Compiler.Service (>= 0.0.87)
     NuGet.CommandLine (2.8.6)
-    NUnit (2.6.4)
-    NUnit.Runners (2.6.4)
-    SourceLink.Fake (1.0.0)
+    NUnit (3.0.0-beta-4)
+    NUnit.Console (3.0.0-beta-4)
+    NUnitTestAdapter (2.0.0)
+    SourceLink.Fake (1.1.0)
```
Comment (kurtschelfthout) [inline paket.lock]: SourceLink, ditto.

File: src/FsCheck.NUnit/FsCheckPropertyAttribute.fs [hunk 7]
```diff
@@ -39,3 +77,101 @@ type PropertyAttribute() =
...
+            testResult.SetResult(ResultState(TestStatus.Passed))
+        | TestResult.Exhausted testdata ->
+            let msg = sprintf "Exhausted: %s" (Runner.onFinishedToString "" testRunner.Result)
+            testResult.SetResult(new ResultState(TestStatus.Failed, msg))
+        | TestResult.False (testdata, originalArgs, shrunkArgs, Outcome.Exception e, seed)  ->
+            let msg = sprintf "%s" (Runner.onFailureToString "" testdata originalArgs shrunkArgs seed)
+            testResult.SetResult(new ResultState(TestStatus.Failed, msg))
+        | TestResult.False (testdata, originalArgs, shrunkArgs, outcome, seed) ->
+            let msg = sprintf "%s" (Runner.onFinishedToString "" testRunner.Result)
+            testResult.SetResult(new ResultState(TestStatus.Failed, msg))
```
Comment (kurtschelfthout) [inline src/FsCheck.NUnit/FsCheckPropertyAttribute.fs]: I'll take your word for it on this one... :)

File: src/FsCheck.Xunit/FsCheck.Xunit.fsproj [hunk 8]
```diff
@@ -87,7 +87,7 @@
 	-->
   <Import Project="..\..\.paket\paket.targets" />
   <Choose>
-    <When Condition="($(TargetFrameworkIdentifier) == '.NETFramework' And ($(TargetFrameworkVersion) == 'v2.0' Or $(TargetFrameworkVersion) == 'v3.0' Or $(TargetFrameworkVersion) == 'v3.5' Or $(TargetFrameworkVersion) == 'v4.0' Or $(TargetFrameworkVersion) == 'v4.5' Or $(TargetFrameworkVersion) == 'v4.5.1' Or $(TargetFrameworkVersion) == 'v4.5.2' Or $(TargetFrameworkVersion) == 'v4.5.3' Or $(TargetFrameworkVersion) == 'v4.6')) Or ($(TargetFrameworkIdentifier) == 'MonoAndroid') Or ($(TargetFrameworkIdentifier) == 'MonoTouch')">
+    <When Condition="$(TargetFrameworkIdentifier) == '.NETFramework' And ($(TargetFrameworkVersion) == 'v2.0' Or $(TargetFrameworkVersion) == 'v3.0' Or $(TargetFrameworkVersion) == 'v3.5' Or $(TargetFrameworkVersion) == 'v4.0' Or $(TargetFrameworkVersion) == 'v4.5' Or $(TargetFrameworkVersion) == 'v4.5.1' Or $(TargetFrameworkVersion) == 'v4.5.2' Or $(TargetFrameworkVersion) == 'v4.5.3' Or $(TargetFrameworkVersion) == 'v4.6')">
```
Comment (kurtschelfthout) [inline src/FsCheck.Xunit/FsCheck.Xunit.fsproj]: Unclear where this change comes from...?

File: tests/FsCheck.Test/FsCheck.Test.fsproj [hunk 9]
```diff
@@ -114,18 +120,18 @@
...
+    <When Condition="($(TargetFrameworkIdentifier) == 'WindowsPhone' And ($(TargetFrameworkVersion) == 'v8.0' Or $(TargetFrameworkVersion) == 'v8.1')) Or ($(TargetFrameworkIdentifier) == 'MonoAndroid') Or ($(TargetFrameworkIdentifier) == 'MonoTouch') Or ($(TargetFrameworkIdentifier) == 'Xamarin.iOS') Or ($(TargetFrameworkIdentifier) == 'Xamarin.Mac') Or ($(TargetFrameworkProfile) == 'Profile7') Or ($(TargetFrameworkProfile) == 'Profile31') Or ($(TargetFrameworkProfile) == 'Profile32') Or ($(TargetFrameworkProfile) == 'Profile44') Or ($(TargetFrameworkProfile) == 'Profile49') Or ($(TargetFrameworkProfile) == 'Profile78') Or ($(TargetFrameworkProfile) == 'Profile84') Or ($(TargetFrameworkProfile) == 'Profile111') Or ($(TargetFrameworkProfile) == 'Profile151') Or ($(TargetFrameworkProfile) == 'Profile157') Or ($(TargetFrameworkProfile) == 'Profile259')">
       <PropertyGroup>
         <__paket__xunit_runner_visualstudio_props>portable-net45+aspnetcore50+win+wpa81+wp80+monotouch+monoandroid\xunit.runner.visualstudio</__paket__xunit_runner_visualstudio_props>
       </PropertyGroup>
```
Comment (kurtschelfthout) [inline tests/FsCheck.Test/FsCheck.Test.fsproj]: Is this a re-ordering of some kind?

File: tests/FsCheck.Test/FsCheck.Test.fsproj [hunk 10]
```diff
@@ -96,8 +96,14 @@
 	</Target>
 	-->
   <Import Project="..\..\.paket\paket.targets" />
+  <Target Name="BeforeBuild">
+    <ItemGroup>
+      <xunitrunner Include="$(ProjectDir)..\..\packages\xunit.runner.visualstudio\**\*.*" />
+    </ItemGroup>
+    <Copy SourceFiles="@(xunitrunner)" DestinationFolder="$(ProjectDir)..\..\packages\xunit.runner.visualstudio.2.0.0\%(RecursiveDir)" SkipUnchangedFiles="true" />
+  </Target>
```
Comment (kurtschelfthout) [inline tests/FsCheck.Test/FsCheck.Test.fsproj]: This target should no longer be necessary with the version of paket; it looks like this is a merge problem.

**Comment by kurtschelfthout** [comment]
Very nice, loved to see FsCheck.NUnit.Addin go. Bunch of comments inline.

**Comment by jhamm** [comment]
Thanks for the feedback! I will work on the changes over the weekend.

**Comment by jhamm** [comment]
The changes you have suggested have been applied. The diff looks much cleaner now. Thanks!

There were a few other small things that were modified as well.

Updated the Documentation to remove the note about the Temporary fix to get NUnit.Addin working for your test.
The Documentation_ja.md still needs this modification.

Removed references to the FsCheck.NUnit.Addin from the mono solution, build.fsx, and FsCheck.Nunit.Examples project

Modified FsCheck.NUnit.nuspec:
to remove install.ps1 and FsCheckAddin.fs because they are no longer needed.
to not reference FsCheck.NUnit.Addin
remove NUnit.Runners from the dependency list

**Comment by kurtschelfthout** [comment]
This is merged on branch `nunit3`. Thanks a bunch for all of this! I'll monitor the AppVeyor build and update #140 with the build number so people can test.
